File: game_bot/game_agent.py
"""
GameAgent – the top-level orchestrator.

The agent runs a continuous perception-decision-action loop:

  ┌──────────────────────────────────────────────────────────┐
  │  PERCEIVE  →  DECIDE  →  ACT  →  EVALUATE  →  LEARN     │
  │      ↑                                          │        │
  │      └──────────────────────────────────────────┘        │
  └──────────────────────────────────────────────────────────┘

Perceive  – take a screenshot, run VisionAnalyzer
Decide    – pick the best next action (vision suggestion + strategy memory)
Act       – dispatch via ActionExecutor
Evaluate  – classify outcome, detect achievements, check if stuck
Learn     – periodic reflection via LearningEngine; extract tutorial knowledge

The agent models all aspects of a human gamer:
  • Completes tutorials by following on-screen instructions
  • Adapts strategy based on past successes and failures
  • Explores when stuck rather than hammering the same button
  • Tracks achievements and progress
  • Knows when it truly cannot progress and stops gracefully
"""

import logging
import random
import time
from dataclasses import dataclass
from typing import Optional

import anthropic

from .action_executor import (
    OUTCOME_FAILURE,
    OUTCOME_NO_CHANGE,
    OUTCOME_PROGRESS,
    OUTCOME_SUCCESS,
    ActionExecutor,
)
from .config import config
from .learning_engine import LearningEngine
from .memory_system import Achievement, MemorySystem
from .mobile_controller import MobileController, ScreenState
from .vision_analyzer import ScreenAnalysis, VisionAnalyzer

log = logging.getLogger(__name__)

# How many consecutive perception cycles with "same screen" before declaring victory
_GAME_COMPLETE_THRESHOLD = 30
# After this many stuck-recovery attempts with zero progress we stop
_MAX_RECOVERY_ATTEMPTS = 5


@dataclass
class AgentStatus:
    running: bool
    total_actions: int
    session_id: int
    last_screen_type: str
    last_outcome: str
    consecutive_stuck: int
    achievements: int


class GameAgent:
    """
    Autonomous game-playing agent.

    Usage:
        agent = GameAgent()
        agent.run()
    """

    def __init__(self):
        log.info("Initialising GameAgent …")
        self._memory   = MemorySystem()
        self._ctrl     = MobileController()
        self._analyzer = VisionAnalyzer()
        self._executor = ActionExecutor(self._ctrl, self._analyzer, self._memory)
        self._learner  = LearningEngine(self._memory)
        self._client   = anthropic.Anthropic(api_key=config.anthropic_api_key)

        self._session_id: int = 0
        self._running: bool = False
        self._recovery_attempts: int = 0
        self._last_analysis: Optional[ScreenAnalysis] = None
        self._prev_screen: Optional[ScreenState] = None
        self._same_screen_count: int = 0

    # ── Public API ────────────────────────────────────────────────────────────

    def run(self) -> None:
        """Main entry point – blocks until the game is complete or the limit is hit."""
        self._session_id = self._memory.start_session()
        self._running = True
        log.info("Session %d started", self._session_id)

        # Ensure game is running
        if config.game_package and not self._ctrl.is_game_running():
            log.info("Launching game …")
            self._ctrl.launch_game()
            time.sleep(4)

        try:
            self._loop()
        except KeyboardInterrupt:
            log.info("Interrupted by user")
        except Exception as e:
            log.exception("Fatal error in agent loop: %s", e)
        finally:
            self._shutdown()

    def status(self) -> AgentStatus:
        return AgentStatus(
            running=self._running,
            total_actions=self._memory.action_count(),
            session_id=self._session_id,
            last_screen_type=self._last_analysis.screen_type if self._last_analysis else "–",
            last_outcome="–",
            consecutive_stuck=self._executor.consecutive_no_change,
            achievements=len(self._memory.get_achievements()),
        )

    # ── Core loop ─────────────────────────────────────────────────────────────

    def _loop(self) -> None:
        while self._running:
            total = self._memory.action_count()

            if total >= config.max_session_actions:
                log.info("Reached max session actions (%d). Stopping.", total)
                break

            # ── 1. PERCEIVE ──────────────────────────────────────────────
            screen = self._ctrl.capture_screen()
            time.sleep(config.screenshot_interval * 0.3)  # brief stabilise

            # Build context string for the vision call
            context = self._build_perception_context()
            analysis = self._analyzer.analyse(screen, context)
            self._last_analysis = analysis

            log.info(
                "[%d] %-12s | %s",
                total, analysis.screen_type, analysis.objective[:70],
            )

            # ── 2. EXTRACT TUTORIAL KNOWLEDGE ────────────────────────────
            if analysis.screen_type == "tutorial":
                self._learner.extract_tutorial_knowledge(analysis)

            # ── 3. CHECK FOR GAME COMPLETION ─────────────────────────────
            if self._is_game_complete(screen, analysis):
                log.info("Game appears to be fully complete. Stopping.")
                break

            # ── 4. CHECK IF STUCK ────────────────────────────────────────
            if self._memory.is_stuck():
                log.warning("Bot is STUCK – triggering recovery")
                recovered = self._recover(screen, analysis)
                if not recovered:
                    self._recovery_attempts += 1
                    if self._recovery_attempts >= _MAX_RECOVERY_ATTEMPTS:
                        log.info(
                            "Cannot progress after %d recovery attempts. "
                            "Game may be at its limit.",
                            _MAX_RECOVERY_ATTEMPTS,
                        )
                        break
                else:
                    self._recovery_attempts = 0
                continue

            # ── 5. DECIDE ACTION ─────────────────────────────────────────
            action = self._decide(analysis)

            # ── 6. ACT ───────────────────────────────────────────────────
            result = self._executor.execute_with_retry(
                action, screen, analysis,
                max_retries=config.max_retry_same_action,
            )

            # ── 7. EVALUATE & DETECT ACHIEVEMENTS ────────────────────────
            if result.outcome in (OUTCOME_PROGRESS, OUTCOME_SUCCESS):
                # Need the post-action screen analysis to detect achievements
                post_analysis = self._analyzer.analyse(result.screen_after)
                achievements = self._learner.detect_achievement(
                    analysis, post_analysis, self._memory.action_count()
                )
                for ach in achievements:
                    is_new = self._memory.record_achievement(ach)
                    if is_new:
                        log.info("🏆  %s – %s", ach.title, ach.description)
                self._last_analysis = post_analysis

            # ── 8. PERIODIC REFLECTION ────────────────────────────────────
            if self._learner.should_reflect():
                log.info("Triggering reflection …")
                refl = self._learner.reflect(analysis)
                if refl.failure_patterns:
                    log.info("Failure patterns: %s", "; ".join(refl.failure_patterns))
                log.info("Learned: %s", refl.summary)

            self._prev_screen = result.screen_after
            time.sleep(config.screenshot_interval * 0.5)

    # ── Decision making ───────────────────────────────────────────────────────

    def _decide(self, analysis: ScreenAnalysis) -> dict:
        """
        Choose the next action.

        Priority order:
          1. If screen is a tutorial  →  follow the tutorial instruction
          2. Check memory for a proven strategy for this situation
          3. Use VisionAnalyzer's suggestion (Claude already reasoned about it)
          4. Exploration: occasionally try something novel
        """
        total = self._memory.action_count()

        # 1. Tutorial mode: always follow the tutorial
        if analysis.screen_type == "tutorial" and analysis.tutorial_instruction:
            return self._decide_tutorial(analysis)

        # 2. Known strategy for this screen type
        strategies = self._memory.get_strategies(analysis.screen_type)
        if strategies and random.random() > config.exploration_rate:
            best = max(strategies, key=lambda s: s.success_rate * (s.success_count + 1))
            if best.success_rate > 0.5 and best.success_count >= 2:
                log.debug("Using strategy: %s (%.0f%%)", best.description, best.success_rate * 100)
                return {
                    "type": best.action_type,
                    "nx": best.nx,
                    "ny": best.ny,
                    "params": best.params,
                    "reasoning": f"[Strategy] {best.description}",
                }

        # 3. Vision-suggested action (default path)
        action = {
            "type": analysis.suggested_action_type,
            "nx": analysis.suggested_nx,
            "ny": analysis.suggested_ny,
            "params": analysis.suggested_params,
            "reasoning": analysis.suggested_reasoning,
        }

        # 4. Exploration jitter – slightly perturb coordinates to find new elements
        if random.random() < config.exploration_rate:
            action = self._exploration_action(analysis)

        return action

    def _decide_tutorial(self, analysis: ScreenAnalysis) -> dict:
        """
        In tutorial screens, the bot must follow on-screen instructions precisely.
        The vision analyzer has already extracted the instruction and suggested
        where to tap; we trust it here.
        """
        # If there's an arrow or highlighted element, tap the highest-confidence one
        if analysis.interactive_elements:
            best = max(analysis.interactive_elements, key=lambda e: e.confidence)
            return {
                "type": "tap",
                "nx": best.nx,
                "ny": best.ny,
                "params": {},
                "reasoning": f"[Tutorial] Follow instruction: {analysis.tutorial_instruction[:60]}",
            }
        return {
            "type": analysis.suggested_action_type,
            "nx": analysis.suggested_nx,
            "ny": analysis.suggested_ny,
            "params": analysis.suggested_params,
            "reasoning": f"[Tutorial] {analysis.suggested_reasoning}",
        }

    def _exploration_action(self, analysis: ScreenAnalysis) -> dict:
        """
        Pick a random interactive element to explore, or nudge the
        vision-suggested coordinates slightly to discover adjacent elements.
        """
        if analysis.interactive_elements:
            chosen = random.choice(analysis.interactive_elements)
            return {
                "type": "tap",
                "nx": chosen.nx,
                "ny": chosen.ny,
                "params": {},
                "reasoning": f"[Explore] Trying element: {chosen.label}",
            }
        # Gaussian jitter around the suggested point
        nx = min(0.95, max(0.05, analysis.suggested_nx + random.gauss(0, 0.1)))
        ny = min(0.95, max(0.05, analysis.suggested_ny + random.gauss(0, 0.1)))
        return {
            "type": "tap",
            "nx": nx,
            "ny": ny,
            "params": {},
            "reasoning": "[Explore] Jittered tap for discovery",
        }

    # ── Stuck recovery ────────────────────────────────────────────────────────

    def _recover(self, screen: ScreenState, analysis: ScreenAnalysis) -> bool:
        """
        Execute a sequence of recovery actions proposed by the LearningEngine.
        Returns True if at least one action produced a visible change.
        """
        recovery_actions = self._learner.build_recovery_plan(analysis)
        self._executor.reset_no_change_counter()
        made_progress = False

        for action in recovery_actions:
            current_screen = self._ctrl.capture_screen()
            result = self._executor.execute(action, current_screen, analysis)
            if result.outcome in (OUTCOME_SUCCESS, OUTCOME_PROGRESS):
                made_progress = True
                break
            time.sleep(1.0)

        return made_progress

    # ── Game completion detection ─────────────────────────────────────────────

    def _is_game_complete(
        self, screen: ScreenState, analysis: ScreenAnalysis
    ) -> bool:
        """
        Heuristic: game is "complete" when either:
          - The screen has been identical for many consecutive cycles, AND
          - We are stuck (recovery exhausted), OR
          - The analysis objective suggests there is nothing left to do.
        """
        nothing_to_do_phrases = [
            "nothing to do", "no more levels", "game complete",
            "fully upgraded", "max level", "all content cleared",
        ]
        obj_lower = analysis.objective.lower()
        if any(phrase in obj_lower for phrase in nothing_to_do_phrases):
            return True

        if self._prev_screen is not None:
            if self._analyzer.is_same_screen(self._prev_screen, screen):
                self._same_screen_count += 1
            else:
                self._same_screen_count = 0

        if self._same_screen_count >= _GAME_COMPLETE_THRESHOLD:
            return True

        return False

    # ── Context builder ───────────────────────────────────────────────────────

    def _build_perception_context(self) -> str:
        """
        Assemble a compact text context to include in vision analysis calls.
        Helps Claude make decisions that are consistent with history.
        """
        parts = [
            "=== RECENT ACTIONS ===",
            self._memory.recent_actions_summary(10),
            "",
            "=== GAME KNOWLEDGE ===",
            self._memory.knowledge_summary(20),
            "",
            "=== ACHIEVEMENTS ===",
            self._memory.achievements_summary(),
        ]
        return "\n".join(parts)

    # ── Shutdown ──────────────────────────────────────────────────────────────

    def _shutdown(self) -> None:
        self._running = False
        total = self._memory.action_count()
        achievements = self._memory.get_achievements()
        notes = (
            f"Session ended. Total actions: {total}. "
            f"Achievements: {len(achievements)}."
        )
        self._memory.end_session(self._session_id, notes)

        log.info("=" * 60)
        log.info("Session complete. Total actions: %d", total)
        log.info("Achievements earned:")
        for ach in achievements:
            log.info("  ✓ %s – %s", ach.title, ach.description)
        log.info("=" * 60)

File: android-game-bot/app/src/main/java/com/gamebot/agent/GameAgent.kt
package com.gamebot.agent

import android.util.Log
import com.gamebot.data.db.ActionEntity
import com.gamebot.data.db.AchievementEntity
import com.gamebot.data.repository.MemoryRepository
import com.gamebot.model.ScreenAnalysis
import com.gamebot.service.GameBotAccessibilityService
import com.gamebot.service.ScreenCaptureService
import kotlinx.coroutines.delay
import kotlin.random.Random

private const val TAG = "GameAgent"
private const val SCREENSHOT_INTERVAL_MS  = 2_000L
private const val SAME_SCREEN_STOP        = 30
private const val MAX_RECOVERY_ATTEMPTS   = 5
private const val MAX_SESSION_ACTIONS     = 5_000

/**
 * Top-level orchestrator — the full perceive → plan → decide → act → evaluate → learn loop.
 *
 * Architecture after integrating SIMA-2 insights:
 *
 *  ┌─────────────────────────────────────────────────────────────────────┐
 *  │  SLOW LOOP (every ~12 fast steps) — HierarchicalPlanner            │
 *  │    claude-opus-4-7  →  high-level goal + updated memory summary    │
 *  │    RewardModel scores the previous goal                            │
 *  │    Self-task generation when goal is complete                      │
 *  ├─────────────────────────────────────────────────────────────────────┤
 *  │  FAST LOOP (every ~2 seconds) — VisionAnalyzer + SkillsLibrary     │
 *  │    claude-sonnet-4-6  →  single concrete action                    │
 *  │    Priority: tutorial → skill → strategy → vision suggestion       │
 *  └─────────────────────────────────────────────────────────────────────┘
 *
 *  RecurrentMemory  — rolling text summary passed to every LLM call
 *  LearningEngine   — periodic reflection + tutorial extraction
 *  SkillsLibrary    — pre-seeded + learned reusable atomic skills
 */
class GameAgent(
    private val screenCapture: ScreenCaptureService,
    private val accessibility: GameBotAccessibilityService,
    private val visionAnalyzer: VisionAnalyzer,
    private val memory: MemoryRepository,
    private val onStatusUpdate: (String) -> Unit = {},
    private val onComplete: () -> Unit = {},
    private val explorationRate: Float = 0.12f,
) {
    private val executor   = ActionExecutor(accessibility, screenCapture, visionAnalyzer)
    private val learner    = LearningEngine(memory, memory.getApiKeySync())
    private val planner    = HierarchicalPlanner(memory.getApiKeySync())
    private val skills     = SkillsLibrary(memory)
    private val recMem     = RecurrentMemory()

    private var sameScreenCount  = 0
    private var prevImageBase64  = ""
    private var recoveryAttempts = 0
    private var running          = false
    private var stepCount        = 0

    // ── Entry point ───────────────────────────────────────────────────────────

    suspend fun run() {
        running = true
        memory.startSession()
        skills.bootstrap()
        recMem.reset()

        // Initial slow-loop planning before any action
        val initScreen = screenCapture.captureBase64()
        if (initScreen != null) {
            planner.triggerSlowLoop(initScreen, "Session start", memory.knowledgeSummary(15))
            recMem.update(planner.rollingSummary)
            recMem.currentGoal = planner.currentGoal
        }

        while (running && memory.actionCount() < MAX_SESSION_ACTIONS) {
            stepCount++

            // ── 1. PERCEIVE ──────────────────────────────────────────────────
            val imageBase64 = screenCapture.captureBase64() ?: run { delay(500); continue }

            // Build rich context from recurrent memory + DB
            val context = buildString {
                append(recMem.toPromptContext())
                append("\n=== CURRENT HIGH-LEVEL GOAL ===\n")
                append(planner.currentGoal)
                append("\n\n=== RECENT ACTIONS ===\n")
                append(memory.recentActionsSummary(8))
            }

            val analysis = visionAnalyzer.analyse(imageBase64, context)
            val statusLine = "[${memory.actionCount()}] ${analysis.screenType}: ${analysis.objective.take(55)}"
            onStatusUpdate(statusLine)
            Log.i(TAG, statusLine)

            // ── 2. SLOW PLANNING TICK (updates goal + memory summary) ────────
            val highLevelGoal = planner.tick(
                imageBase64,
                memory.recentActionsSummary(10),
                memory.knowledgeSummary(15),
            )
            recMem.update(planner.rollingSummary)
            recMem.currentGoal    = planner.currentGoal
            recMem.lastNarration  = planner.currentNarration

            // ── 3. TUTORIAL KNOWLEDGE EXTRACTION ────────────────────────────
            if (analysis.screenType == "tutorial") learner.extractTutorialKnowledge(analysis)

            // ── 4. GAME-COMPLETE CHECK ────────────────────────────────────────
            if (isGameComplete(imageBase64, analysis)) {
                Log.i(TAG, "Game appears complete – stopping")
                onStatusUpdate("Game complete!")
                recMem.addHighlight("Game completed after ${memory.actionCount()} actions")
                break
            }

            // ── 5. STUCK CHECK + RECOVERY ────────────────────────────────────
            if (memory.isStuck()) {
                Log.w(TAG, "STUCK – attempt ${recoveryAttempts + 1}")
                onStatusUpdate("Stuck… trying recovery plan")
                val recovered = recover(imageBase64, analysis)
                if (!recovered) {
                    recoveryAttempts++
                    if (recoveryAttempts >= MAX_RECOVERY_ATTEMPTS) {
                        Log.i(TAG, "Cannot progress – stopping")
                        onStatusUpdate("Cannot progress further")
                        break
                    }
                } else {
                    recoveryAttempts = 0
                }
                continue
            }

            // ── 6. DECIDE ACTION (fast loop) ──────────────────────────────────
            val action = decide(analysis, highLevelGoal)

            // ── 7. ACT ────────────────────────────────────────────────────────
            val result = executor.execute(
                type        = action["type"] ?: "tap",
                nx          = action["nx"]?.toFloatOrNull() ?: 0.5f,
                ny          = action["ny"]?.toFloatOrNull() ?: 0.5f,
                params      = action.filter { it.key !in listOf("type","nx","ny","reasoning") },
                imageBefore = imageBase64,
            )

            // ── 8. LOG ACTION ─────────────────────────────────────────────────
            memory.logAction(ActionEntity(
                screenType = analysis.screenType,
                actionType = result.actionType,
                nx         = result.nx,
                ny         = result.ny,
                reasoning  = action["reasoning"] ?: "",
                outcome    = result.outcome.name.lowercase(),
            ))

            // ── 9. DETECT ACHIEVEMENTS ────────────────────────────────────────
            if (result.imageAfterBase64 != null) {
                val post = visionAnalyzer.analyse(result.imageAfterBase64)
                detectAchievements(analysis, post)

                // Score goal completion every ~50 actions
                if (stepCount % 50 == 0) {
                    val score = planner.scoreGoalAttempt(
                        highLevelGoal, result.imageAfterBase64,
                        memory.recentActionsSummary(20),
                    )
                    recMem.recordGoalScore(highLevelGoal, score)
                    // If goal well-achieved, generate a fresh self-improvement task
                    if (score >= 7f) {
                        val nextTask = planner.generateNextTask(
                            result.imageAfterBase64,
                            memory.achievementsSummary(),
                            memory.knowledgeSummary(15),
                        )
                        recMem.addHighlight("New self-task: $nextTask")
                        Log.i(TAG, "Self-generated next task: $nextTask")
                    }
                }
            }

            // ── 10. PERIODIC DEEP REFLECTION ─────────────────────────────────
            if (learner.shouldReflect()) {
                onStatusUpdate("Reflecting on what I've learned…")
                val summary = learner.reflect(analysis)
                if (summary.isNotBlank()) recMem.addHighlight("Reflection: $summary")
            }

            prevImageBase64 = imageBase64
            delay(SCREENSHOT_INTERVAL_MS / 2)
        }

        memory.endSession()
        onComplete()
    }

    fun stop() { running = false }

    // ── Decision making (fast loop) ───────────────────────────────────────────

    /**
     * Priority chain:
     *   1. Tutorial mode  → follow on-screen instruction
     *   2. Skill match    → use pre-seeded / learned atomic skill
     *   3. Known strategy → highest success-rate strategy for this screen
     *   4. Vision model   → claude-sonnet-4-6 suggestion (default)
     *   5. Exploration    → random element from detected interactive list
     */
    private fun decide(analysis: ScreenAnalysis, highLevelGoal: String): Map<String, String> {

        // 1. Tutorial
        if (analysis.screenType == "tutorial") return decideTutorial(analysis)

        // 2. Skill match from SkillsLibrary
        val skill = skills.suggestSkill(analysis.screenType, analysis.objective, analysis.blockingIssues)
        if (skill != null) {
            Log.d(TAG, "SKILL match: ${skill.situation}")
            return mapOf(
                "type"      to skill.actionType,
                "nx"        to skill.nx.toString(),
                "ny"        to skill.ny.toString(),
                "reasoning" to "[Skill] ${skill.description}",
            )
        }

        // 3. High-confidence learned strategy
        val strategy = memory.getBestStrategy(analysis.screenType)
        if (strategy != null && strategy.successCount >= 2 &&
            strategy.successRate > 0.5f && Random.nextFloat() > explorationRate) {
            return mapOf(
                "type"      to strategy.actionType,
                "nx"        to strategy.nx.toString(),
                "ny"        to strategy.ny.toString(),
                "reasoning" to "[Strategy] ${strategy.description}",
            )
        }

        // 4. Vision-suggested action (default fast loop)
        val sa = analysis.suggestedAction
        val base = mapOf(
            "type"      to sa.type,
            "nx"        to sa.nx.toString(),
            "ny"        to sa.ny.toString(),
            "reasoning" to sa.reasoning,
        ) + sa.params

        // 5. Exploration jitter
        if (Random.nextFloat() < explorationRate && analysis.interactiveElements.isNotEmpty()) {
            val el = analysis.interactiveElements.random()
            return mapOf(
                "type"      to "tap",
                "nx"        to el.nx.toString(),
                "ny"        to el.ny.toString(),
                "reasoning" to "[Explore] ${el.label}",
            )
        }

        return base
    }

    private fun decideTutorial(analysis: ScreenAnalysis): Map<String, String> {
        val best = analysis.interactiveElements.maxByOrNull { it.confidence }
        val sa   = analysis.suggestedAction
        return if (best != null) mapOf(
            "type" to "tap", "nx" to best.nx.toString(), "ny" to best.ny.toString(),
            "reasoning" to "[Tutorial] ${analysis.tutorialInstruction.take(60)}",
        ) else mapOf(
            "type" to sa.type, "nx" to sa.nx.toString(), "ny" to sa.ny.toString(),
            "reasoning" to "[Tutorial] ${sa.reasoning}",
        )
    }

    // ── Recovery ──────────────────────────────────────────────────────────────

    private suspend fun recover(imageBase64: String, analysis: ScreenAnalysis): Boolean {
        executor.resetNoChangeCounter()
        val actions = learner.buildRecoveryPlan(analysis)
        for (action in actions) {
            val img = screenCapture.captureBase64() ?: continue
            val result = executor.execute(
                type  = action["type"] ?: "tap",
                nx    = action["nx"]?.toFloatOrNull() ?: 0.5f,
                ny    = action["ny"]?.toFloatOrNull() ?: 0.5f,
                imageBefore = img,
            )
            if (result.outcome == Outcome.SUCCESS || result.outcome == Outcome.PROGRESS) return true
            delay(1_000L)
        }
        return false
    }

    // ── Achievement detection ─────────────────────────────────────────────────

    private fun detectAchievements(before: ScreenAnalysis, after: ScreenAnalysis) {
        val count = memory.actionCount()
        if (before.screenType == "tutorial" && after.screenType != "tutorial") {
            memory.recordAchievement(AchievementEntity(
                title = "Tutorial Completed", description = "Moved to ${after.screenType}",
                screenType = after.screenType, actionCount = count,
            ))
            recMem.addHighlight("Tutorial completed!")
        }
        if (after.screenType == "victory") {
            val level = after.gameState.level ?: before.gameState.level
            memory.recordAchievement(AchievementEntity(
                title = if (level != null) "Level $level Cleared" else "Level Cleared",
                description = after.objective, screenType = "victory", actionCount = count,
            ))
            recMem.addHighlight("Level ${level ?: "?"} cleared!")
        }
        listOf("level" to after.gameState.level, "score" to after.gameState.score?.toInt()).forEach { (key, newVal) ->
            val oldVal = when (key) {
                "level" -> before.gameState.level
                else    -> before.gameState.score?.toInt()
            }
            if (oldVal != null && newVal != null && newVal > oldVal) {
                memory.recordAchievement(AchievementEntity(
                    title = "${key.replaceFirstChar { it.uppercase() }} Increased",
                    description = "$key: $oldVal → $newVal",
                    screenType = after.screenType, actionCount = count,
                ))
            }
        }
    }

    // ── Game-complete detection ───────────────────────────────────────────────

    private suspend fun isGameComplete(imageBase64: String, analysis: ScreenAnalysis): Boolean {
        val done = listOf(
            "nothing to do", "no more levels", "game complete",
            "max level", "fully upgraded", "all content cleared",
        ).any { it in analysis.objective.lowercase() }
        if (done) return true
        if (prevImageBase64.isNotEmpty()) {
            if (visionAnalyzer.isSameScreen(prevImageBase64, imageBase64)) sameScreenCount++
            else sameScreenCount = 0
        }
        return sameScreenCount >= SAME_SCREEN_STOP
    }

    // ── Expose recurrent memory for UI ────────────────────────────────────────

    fun getRecurrentMemory(): RecurrentMemory = recMem

    private suspend fun MemoryRepository.knowledgeSummary(n: Int): String = buildString {
        val items = try {
            val field = this@knowledgeSummary::class.java.getDeclaredMethod("knowledgeSummary", Int::class.java)
            field.invoke(this@knowledgeSummary, n) as? String ?: ""
        } catch (e: Exception) { "" }
        append(items)
    }

    private suspend fun MemoryRepository.achievementsSummary(): String = runCatching {
        val m = this::class.java.getDeclaredMethod("achievementsSummary")
        m.invoke(this) as? String ?: ""
    }.getOrDefault("")
}
